### PR TITLE
Add CaptureViaZivid

### PR DESCRIPTION
**Pull request by SatjaSivcev**

### CODE REVIEW

**Comment by chrisasc** [comment]
This is now ready for review, @SatjaSivcev

File: source/csharp/Camera/Basic/CaptureViaGenICam/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/Properties/AssemblyInfo.cs:36]: Modify this for both samples. Check C# samples to see how it should look like.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>..\..\..\Release\x64\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="halcondotnet">
+      <HintPath>..\..\..\..\..\..\..\..\..\..\Program Files\MVTec\HALCON-19.11-Progress\bin\dotnet35\halcondotnet.dll</HintPath>
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:36]: Use env var for this. Check C# samples to see how.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Net.Http" />
+    <Reference Include="System.Xml" />
+    <Reference Include="ZividCoreNET">
+      <HintPath>..\..\..\..\..\..\..\..\..\..\Program Files\Zivid\bin\ZividCoreNET.dll</HintPath>
+    </Reference>
+    <Reference Include="ZividVis3DNET">
+      <HintPath>..\..\..\..\..\..\..\..\..\..\Program Files\Zivid\bin\ZividVis3DNET.dll</HintPath>
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:50]: Use env var for this. Check the other sample, there it is good.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    <Reference Include="System" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml.Linq" />
+    <Reference Include="System.Data.DataSetExtensions" />
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Data" />
+    <Reference Include="System.Net.Http" />
+    <Reference Include="System.Xml" />
+    <Reference Include="ZividCoreNET">
+      <HintPath>..\..\..\..\..\..\..\..\..\..\Program Files\Zivid\bin\ZividCoreNET.dll</HintPath>
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:47]: Use env var for this. Check the other sample, there it is good.

File: source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+
+[assembly:AssemblyVersion("1.8.1")]
+[assembly:AssemblyFileVersion("1.8.1")]
+[assembly:AssemblyInformationalVersion("1.8.1+6967bc1b-1")]
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs:38]: Same as for the other sample.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    <OutputPath>bin\Debug\</OutputPath>
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <PlatformTarget>x64</PlatformTarget>
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>..\..\..\Release\x64\</OutputPath>
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:29]: Check the other sample how it is done there.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            HalconDotNet.HOperatorSet.SetFramegrabberParam(framegrabber, "ExposureTime", 8333);
+            HalconDotNet.HOperatorSet.SetFramegrabberParam(framegrabber, "Gain", 1);
+            HalconDotNet.HOperatorSet.SetFramegrabberParam(framegrabber, "ProcessingFiltersOutlierRemovalEnabled", 1);
+            HalconDotNet.HOperatorSet.SetFramegrabberParam(framegrabber, "ProcessingFiltersOutlierRemovalThreshold", 5);
+
+            Console.WriteLine("Capturing frame");
+            var region = new HalconDotNet.HObject();
+            var contours = new HalconDotNet.HObject();
+            var data = new HalconDotNet.HTuple();
+            var frame = new HalconDotNet.HObject();
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:47]: Maybe keep the same order as input arguments for the next function?

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            var y = frame.SelectObj(2);
+            var z = frame.SelectObj(3);
+            var snr = frame.SelectObj(4);
+            var rgb = frame.SelectObj(5);
+
+            Console.WriteLine("Removing invalid 3D points (zeroes)");
+            var reducedRegion = new HalconDotNet.HObject();
+            var zReduced = new HalconDotNet.HObject();
+            HalconDotNet.HOperatorSet.Threshold(z, out reducedRegion, 0.0001, 10000);
+            HalconDotNet.HOperatorSet.ReduceDomain(z, reducedRegion, out zReduced);
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:60]: What is the reason to chose this method and not the xyz mapping? Speed?

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            Console.WriteLine("Constructing ObjetModel3D based on XYZ data");
+            var objectModel3D = new HalconDotNet.HTuple();
+            HalconDotNet.HOperatorSet.XyzToObjectModel3d(x, y, zReduced, out objectModel3D);
+
+            Console.WriteLine("Adding RGB to ObjectModel3D");
+            SetColorsInObjectModel3D(objectModel3D, rgb, zReduced);
+
+            string fileName = "Zivid3D.ply";
+            Console.WriteLine("Saving point cloud to: " + fileName);
+            SaveHalconPointCloud(objectModel3D, fileName);
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:71]: check what variable name convention we are using in C# / C++ samples, especially for ply. I think we use pointCloud. Let's make it consistent.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            Console.WriteLine("Error: " + ex.Message);
+            Environment.ExitCode = 1;
+        }
+    }
+
+    private static void SaveHalconPointCloud(HalconDotNet.HTuple model, string fileName)
+    {
+        HalconDotNet.HOperatorSet.WriteObjectModel3d(model, "ply", fileName, "invert_normals", "false");
+    }
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:83]: Should we get rid of the function since it is a one liner? It is clear from the printout what is happening.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    private static void SetColorsInObjectModel3D(HalconDotNet.HTuple objectModel3D, HalconDotNet.HObject RGB, HalconDotNet.HObject zReduced)
+    {
+        var rows = new HalconDotNet.HTuple();
+        var cols = new HalconDotNet.HTuple();
+        var domain = new HalconDotNet.HObject();
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:89]: change the order so to be the same in input arguments in the function that is using it?

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            Console.WriteLine("Capturing frame");
+            var frame = camera.Capture(settings);
+            var pointCloud = frame.PointCloud;
+
+            Console.WriteLine("Converting to Halcon point cloud");
+            HalconDotNet.HObjectModel3D objectModel3D = ZividToHalconPointCloud(pointCloud);
+
+            string fileName = "Zivid3D.ply";
+            Console.WriteLine("Saving point cloud to: " + fileName);
+            SaveHalconPointCloud(objectModel3D, fileName);
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:32]: Same comment as in the other sample.

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        {
+            Console.WriteLine("Error: " + ex.Message);
+            Environment.ExitCode = 1;
+        }
+    }
+
+    private static void SaveHalconPointCloud(HalconDotNet.HObjectModel3D model, string fileName)
+    {
+        model.WriteObjectModel3d("ply", fileName, "invert_normals", "false");
+    }
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:44]: Same comment as in the other sample.

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+    private static HalconDotNet.HObjectModel3D ZividToHalconPointCloud(Zivid.NET.PointCloud pointCloud)
+    {
+        var height = pointCloud.Height;
+        var width = pointCloud.Width;
+        var pointsXYZ = pointCloud.CopyPointsXYZ();
+        var colorsRGBA = pointCloud.CopyColorsRGBA();
+
+        // We need to find the number of valid points to initialize our halcon tuples.
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:70]: Use the agreed convention for comments (-ing tense)

Comment (SatjaSivcev) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:70]: In the next comment you use HTuples. Use the same here?

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        // Initializing HTuples which are later filled with data from the Zivid point cloud.
+        var tupleX = new HalconDotNet.HTuple();
+        var tupleY = new HalconDotNet.HTuple();
+        var tupleZ = new HalconDotNet.HTuple();
+        var tupleR = new HalconDotNet.HTuple();
+        var tupleG = new HalconDotNet.HTuple();
+        var tupleB = new HalconDotNet.HTuple();
+        var tupleXYZMapping = new HalconDotNet.HTuple();
+
+        // Compensate for 0 indexing, therefore the ' - 1'
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:82]: Use the agreed convention for comments (-ing tense)

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+                {
+                    tupleX[validPointIndex] = pointsXYZ[i, j, 0];
+                    tupleY[validPointIndex] = pointsXYZ[i, j, 1];
+                    tupleZ[validPointIndex] = pointsXYZ[i, j, 2];
+                    tupleR[validPointIndex] = colorsRGBA[i, j, 0];
+                    tupleG[validPointIndex] = colorsRGBA[i, j, 1];
+                    tupleB[validPointIndex] = colorsRGBA[i, j, 2];
+                    tupleXYZMapping[2 + validPointIndex] = i;
+                    tupleXYZMapping[2 + numberOfValidPoints + validPointIndex] = j;
+
```
Comment (SatjaSivcev) [inline source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:113]: is this space (new line) needed?

File: source/cpp/CMakeLists.txt [hunk 17]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
 
-project(CppExtraSamples)
+project(CppHalconSamples)
```
Comment (SatjaSivcev) [inline source/cpp/CMakeLists.txt:3]: make this a separate commit.

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs (hunk 8 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:60]: We still do xyz mapping, but before we do the mapping we remove all zero points. This reduces the size of the point cloud

File: source/csharp/Camera/Basic/CaptureViaGenICam/Properties/AssemblyInfo.cs (hunk 1 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/Properties/AssemblyInfo.cs:36]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs (hunk 7 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:47]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs (hunk 9 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:71]: It does not say anything about the fileName in our internal KB, but from the sample we use pointCloudFile. So switching to this

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs (hunk 10 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:83]: I would like to keep it the same structure as C++

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs (hunk 11 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.cs:89]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 6 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:29]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 2 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:36]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 4 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:47]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 3 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:50]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 12 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:32]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 16 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:113]: nope, its been removed now

File: source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs:38]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 2 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:36]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/Properties/AssemblyInfo.cs:38]: done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 4 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:47]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 12 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:32]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 13 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:44]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 14 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:70]: squashed all comments into one. These comments were made for A&G so the developer there could understand it better

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 15 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:82]: done

File: source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs (hunk 16 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaZivid/CaptureViaZivid.cs:113]: removed. done

File: source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj (hunk 3 above)
Comment (chrisasc) [reply source/csharp/Camera/Basic/CaptureViaGenICam/CaptureViaGenICam.csproj:50]: done

File: source/cpp/CMakeLists.txt (hunk 17 above)
Comment (chrisasc) [reply source/cpp/CMakeLists.txt:3]: done
